README.md
# Пример использования анализатора спектра Satis AF-05 в Python
[![GitHub Workflow Status](https://img.shields.io/github/workflow/status/vb64/Satis_AF05/satis%20pep257?label=Pep257&style=plastic)](https://github.com/vb64/Satis_AF05/actions?query=workflow%3A%22satis+pep257%22)
[![GitHub Workflow Status](https://img.shields.io/github/workflow/status/vb64/Satis_AF05/satis%20tests?label=Python%203.6%203.7%203.8%203.9%203.10&style=plastic)](https://github.com/vb64/Satis_AF05/actions?query=workflow%3A%22satis+tests%22)
[![Codacy Badge](https://app.codacy.com/project/badge/Grade/df385d9214b94778b22c630c7d4baefa)](https://www.codacy.com/gh/vb64/Satis_AF05/dashboard?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=vb64/Satis_AF05&amp;utm_campaign=Badge_Grade)
[![Codacy Badge](https://app.codacy.com/project/badge/Coverage/df385d9214b94778b22c630c7d4baefa)](https://www.codacy.com/gh/vb64/Satis_AF05/dashboard?utm_source=github.com&utm_medium=referral&utm_content=vb64/Satis_AF05&utm_campaign=Badge_Coverage)

Данный пример реализует взаимодействие с анализатором спектра Satis AF-05 в языке программирования Python.
Вы можете использовать готовый исполняемый файл [satis.exe](https://github.com/vb64/Satis_AF05/releases/tag/ver.1.1) или [собрать собственную версию](https://github.com/vb64/Satis_AF05#%D1%81%D0%B1%D0%BE%D1%80%D0%BA%D0%B0-%D1%81%D0%BE%D0%B1%D1%81%D1%82%D0%B2%D0%B5%D0%BD%D0%BD%D0%BE%D0%B9-%D0%B2%D0%B5%D1%80%D1%81%D0%B8%D0%B8-%D0%B8%D1%81%D0%BF%D0%BE%D0%BB%D0%BD%D1%8F%D0%B5%D0%BC%D0%BE%D0%B3%D0%BE-%D1%84%D0%B0%D0%B9%D0%BB%D0%B0-%D0%B8%D0%B7-%D0%B8%D1%81%D1%85%D0%BE%D0%B4%D0%BD%D1%8B%D1%85-%D0%BA%D0%BE%D0%B4%D0%BE%D0%B2) исполняемого файла из исходных кодов.

[![Satis AF-05](/img/device.jpg)](http://www.satis-tl.ru/products/oborudovanie-sistem-upravleniya-i-kontrolya-setey-sputnikovoy-svyazi/analizator-chastotnyy-af-4/)

## Подключение Satis AF-05 к компьютеру

Перед началом работы необходимо подключить анализатор к компьютеру, на котором будет запускаться программа примера.
Нужно соединить кабелем Ethernet анализатор и компьютер напрямую либо через коммутатор.
По умолчанию у анализатора установлен IP адрес 192.168.0.100. Сетевую карту компьютера, соединенную с анализатором, нужно настроить на работу в данной подсети.

Если настройки сделаны правильно и соединение работает, при открытии в браузере адреса 192.168.0.100 вы должны увидеть что-то похожее на следующую картинку.

![Веб-интерфейс](/img/browser.jpg)

## Считывание данных анализатора

Пример реализует два режима работы с анализатором - чтение данных сигнала в заданном диапазоне частот и режим чтения данных для заданной частоты.

### Команда 'read'

Команда 'read' реализует режим чтения данных в заданном диапазоне частот с заданными параметрами.

```bash
satis.exe read --address=192.168.1.100 --freq_start=1400000000 --freq_end=1400330000 --video=50 --rbw=0 --atten=0 --with_data
```

В данном режиме программа принимает следующие параметры

-   --address строка адреса, по которому доступен анализатор. по умолчанию '192.168.0.100'
-   --freq_start начальная частота диапазона в герцах. допустимые значения от 950000000 до 2150000000. по умолчанию 950000000.
-   --freq_end конечная частота диапазона в герцах. допустимые значения от 950000000 до 2150000000. по умолчанию 2150000000.
-   --video частота усреднения результата на периоде времени в герцах. допустимые значения от 0.1 (период 10 секунд) до 6400 (период 0.156 миллисекунды). по умолчанию 100.
-   --rbw разрешение по частоте. целое число от 0 до 4. по умолчанию 2 (400 Гц).
    - 0: 6400 Гц
    - 1: 1600 Гц
    - 2: 400 Гц
    - 3: 100 Гц
    - 4: 25 Гц
-   --atten аттенюация входного сигнала. целое число от 0 до 6. по умолчанию 0 (0 дБ).
    - 0: 0 дБ
    - 1: 5 дБ
    - 2: 10 дБ
    - 3: 16 дБ
    - 4: 21 дБ
    - 5: 26 дБ
    - 6: 31 дБ
-   --with_data вывод на экран значений прочитанных данных. по умолчанию выводится только размер массива считанных данных.

В процессе работы команда выводит на экран данные, получаемые от анализатора. По завершении цикла чтения данных выводится общее количество считанных значений сигнала.

### Команда 'sweep'

Команду 'sweep' можно использовать в ситуации, когда вас интересует значение сигнала на заранее известной вам частоте.
Команда максимально быстро выведет список значений сигнала в окрестностях заданной частоты.

```bash
satis.exe sweep --address=192.168.1.100 --freq_center=1450000000 --datafreq=10 --rbwsweep=0 --atten=0
```

В данном режиме программа принимает следующие параметры

-   --freq_center интересующая частота в Гц. от 950000000 до 2150000000. по умолчанию 950000000.
-   --datafreq период осреднения в Гц. от 1.0 до 20.0. по умолчанию 10.
-   --rbwsweep разрешение по частоте. целое число от 0 до 2. по умолчанию 0 (6400 Гц).
    - 0: 6400 Гц
    - 1: 1600 Гц
    - 2: 400 Гц

Остальные параметры имеют то же значение, что и в команде 'read'.

Команда выведет на экран данные, полученные от анализатора.

## Сборка собственной версии исполняемого файла из исходных кодов

Перечень средств разработки, необходимых для сборки исполняемого файла. Скачать и установить:

-   [Python 3](https://www.python.org/downloads/release/python-3810/)
-   [Git for Windows](https://git-scm.com/download/win) для доступа к репозитарию исходных кодов
-   GNU [Unix Utils](http://unxutils.sourceforge.net/) для операций с makefile

Затем

```bash
git clone git@github.com:vb64/Satis_AF05.git
cd Satis_AF05
make setup PYTHON_BIN=/path/to/python3.exe
make tests
make exe
```

Файл `satis.exe` будет создан в каталоге `dist`.

## Поддержка Сатис
Детальное описание всех режимов работы анализатора можно получить из [документации на устройство](http://satis-tl.ru/documents_support/documents/).

Если у вас возникли трудности при программном доступе к анализатору, вы можете [обратиться в техподдержку](https://www.satis-tl.ru/documents_support/) удобным вам способом или задать вопрос через [форму на сайте производителя](https://satis-tl.ru/contacts/#moscow).


history.txt
22.01.2022 ver.1.1
------------------

- Исправлена ошибка, не позволявшая задать значения 5 (26 дБ) и 6 (31 дБ) для опции `--atten`.

* Исполняемый файл программы пересобран на основе Python 3.7, что позволяет запускать его под Windows7 без дополнительных настроек.

19.01.2022 ver.1.0
------------------

+ Первоначальная версия программы.


tests/test/__init__.py
"""Root class for testing."""
import json
from unittest import TestCase


class MockSocket:
    """Mocked websocket."""

    def __init__(self):
        """Instance."""
        self.addr = None
        self.last_text = None
        self.answers = []

    def connect(self, addr):
        """Connect to address."""
        self.addr = addr

    def send(self, text):
        """Write string to socket."""
        self.last_text = text

    def recv(self):
        """Return string from socket."""
        return json.dumps(self.answers.pop())

    def close(self):
        """Close instance."""
        self.addr = None


class TestBase(TestCase):
    """Base class for tests."""

    def setUp(self):
        """Init tests."""
        TestCase.setUp(self)

        from source.cli import PARSER
        self.options, _args = PARSER.parse_args(args=[])


tests/test/test_satis.py
"""Tests for satis.py.

make test T=test_satis.py
"""
import pytest
from . import TestBase, MockSocket


class TestSatis(TestBase):
    """Tests satis module."""

    @staticmethod
    def test_get_string():
        """Call get_string function."""
        from satis import get_string

        assert get_string([True]) == '[true]'

    @staticmethod
    def test_sweep():
        """Call sweep function."""
        from satis import sweep, Rbw, Attenuation, Key

        socket = MockSocket()
        socket.connect('testsweep')
        socket.answers.append({
          Key.Data: [1, 2],
        })

        assert sweep(socket, 100, Rbw.Hz6400, 100, Attenuation.Db0)

    @staticmethod
    def test_read():
        """Call read function."""
        from satis import read, Rbw, Attenuation, Key, Error

        socket = MockSocket()
        socket.connect('test')
        socket.answers.append({
          Key.First: 3,
          Key.Data: [4, 5],
        })
        socket.answers.append({
          Key.Total: 5,
          Key.First: 0,
          Key.Data: [1, 2, 3],
        })

        assert read(socket, 100, 150, Rbw.Hz6400, 100, Attenuation.Db0, True) > 0

        socket.answers.append({
          Key.First: 0,
          Key.Data: [4, 5],
        })
        socket.answers.append({
          Key.Total: 5,
          Key.First: 0,
          Key.Data: [1, 2, 3],
        })

        with pytest.raises(Error) as exp:
            read(socket, 100, 150, Rbw.Hz6400, 100, Attenuation.Db0, False)
        assert "Error: index " in str(exp)

        socket.close()


tests/test/test_console.py
"""Console client stuff.

make test T=test_console.py
"""
from . import TestBase, MockSocket


class TestConsole(TestBase):
    """Tests console client."""

    def test_wrong(self):
        """Call read command."""
        import cli

        assert cli.main([], self.options) == 1
        assert cli.main(["wrong_command"], self.options) == 1

    def test_sweep(self):
        """Call sweep command."""
        import cli

        saved_websocket = cli.websocket.WebSocket
        saved_sweep = cli.sweep

        cli.websocket.WebSocket = MockSocket
        cli.sweep = lambda socket, start, rbw, vid, att: []

        assert cli.main([cli.Command.Sweep], self.options) == 0

        cli.sweep = saved_sweep
        cli.websocket.WebSocket = saved_websocket

    def test_read(self):
        """Call read command."""
        import cli

        saved_websocket = cli.websocket.WebSocket
        saved_read = cli.read

        cli.websocket.WebSocket = MockSocket
        cli.read = lambda socket, start, end, rbw, vid, att, data: []

        assert cli.main([cli.Command.Read], self.options) == 0

        self.options.with_data = True
        assert cli.main([cli.Command.Read], self.options) == 0

        cli.read = saved_read
        cli.websocket.WebSocket = saved_websocket


source/satis.py
"""Satis AF-5 spectrum analizer stuff."""
import json

MAX_START = 950000000
MAX_END = 2150000000


class Error(Exception):
    """Satis IO exception."""


class Key:
    """Keys for satis json IO."""

    Request = "requestId"
    Start = "fStart"
    End = "fEnd"
    Video = "video"
    Total = "totalFreqPointsInChart"
    First = "frameFirstPointIndex"
    Data = "data"
    Alive = "alive"
    Is_alive = "spectrumAnaliserIsAlive"

    # sweep mode keys
    VideoSweep = "dataFreq"  # from 1 to 20
    Center = "fCentral"  # from MAX_START to MAX_END
    MaxFreq = "maximumFreq"
    MaxVal = "maximumValue"
    Time = "timeStamp"  # milliseconds from last request


class Rbw:
    """Frequency step."""

    name = "rbw"

    Hz6400 = 0
    Hz1600 = 1
    Hz400 = 2
    Hz100 = 3
    Hz25 = 4


class RbwSweep:
    """Frequency step for sweep mode."""

    name = "rbwsweep"

    Hz6400 = 0
    Hz1600 = 1
    Hz400 = 2


class Attenuation:
    """Attenuation of input signal."""

    name = "atten"

    Db0 = 0
    Db5 = 1
    Db10 = 2
    Db16 = 3
    Db21 = 4
    Db26 = 5
    Db31 = 6


class Mode:
    """Operation mode."""

    name = "mode"

    Stop = 0
    Cycle = 1
    Single = 2
    Sweep = 3


CMD_ALIVE = {Key.Alive: True}
CMD_STOP = {Mode.name: Mode.Stop}


def get_string(data):
    """Convert data dict to valid json string."""
    return ''.join(json.dumps(data).split())


def sweep(socket, freq, rbw, video, atten):
    """Return data in sweep mode."""
    socket.send(get_string({
      Key.Request: 6,
      Key.Center: freq,
      Rbw.name: rbw,
      Key.VideoSweep: video,
      Attenuation.name: atten,
      Mode.name: Mode.Sweep,
    }))

    return json.loads(socket.recv())


def dump(result, with_data):
    """Print json data."""
    if not with_data:
        result[Key.Data] = len(result[Key.Data])
    print(result)


def read(socket, freq_start, freq_end, rbw, video, atten, with_data):
    """Return list of satis data."""
    socket.send(get_string({
      Key.Request: 5,
      Key.Start: freq_start,
      Key.End: freq_end,
      Rbw.name: rbw,
      Key.Video: video,
      Attenuation.name: atten,
      Mode.name: Mode.Single,
    }))

    result = json.loads(socket.recv())
    total = result[Key.Total]
    data = result[Key.Data]
    count = len(data)
    dump(result, with_data)

    while count < total:
        result = json.loads(socket.recv())
        if result[Key.First] != count:
            raise Error("Error: index {} count {}". format(result[Key.First], count))

        dump(result, with_data)
        count += len(data)

    return count


source/cli.py
"""Satis AF-5 reader."""
import sys
from optparse import OptionParser  # pylint: disable=deprecated-module
import websocket

from satis import MAX_START, MAX_END, Rbw, RbwSweep, Attenuation, sweep, read

VERSION = '1.1'


class Command:
    """App commands."""

    Read = "read"
    Sweep = "sweep"


def cmd_read(socket, options):
    """App command read."""
    print("Total points:", read(
      socket,
      options.freq_start,
      options.freq_end,
      int(options.rbw),
      options.video,
      int(options.atten),
      options.with_data
    ))


def cmd_sweep(socket, options):
    """App command sweep."""
    print(sweep(
      socket,
      options.freq_center,
      int(options.rbwsweep),
      options.datafreq,
      int(options.atten)
    ))


COMMANDS = {
  Command.Read: cmd_read,
  Command.Sweep: cmd_sweep,
}

RBW = [str(i) for i in [
  Rbw.Hz6400,
  Rbw.Hz1600,
  Rbw.Hz400,
  Rbw.Hz100,
  Rbw.Hz25,
]]

RBW_SWEEP = [str(i) for i in [
  RbwSweep.Hz6400,
  RbwSweep.Hz1600,
  RbwSweep.Hz400,
]]

ATTEN = [str(i) for i in [
  Attenuation.Db0,
  Attenuation.Db5,
  Attenuation.Db10,
  Attenuation.Db16,
  Attenuation.Db21,
  Attenuation.Db26,
  Attenuation.Db31,
]]

PARSER = OptionParser(
  usage="%prog " + ' '.join(["[{}]".format(i) for i in COMMANDS]),
  version="%prog version {}".format(VERSION)
)

PARSER.add_option(
  "--address",
  dest="address",
  default="192.168.0.100",
  help="Set satis websocket address. Default 192.168.0.100"
)
PARSER.add_option(
  "--with_data",
  action="store_true",
  dest="with_data",
  default=False,
  help="Dump received data."
)
PARSER.add_option(
  "--freq_center",
  type="int",
  dest="freq_center",
  default=MAX_START,
  help="Central frequency. Default is {}".format(MAX_START)
)
PARSER.add_option(
  "--freq_start",
  type="int",
  dest="freq_start",
  default=MAX_START,
  help="Start frequency. Default is {}".format(MAX_START)
)
PARSER.add_option(
  "--freq_end",
  type="int",
  dest="freq_end",
  default=MAX_END,
  help="End frequency. Default is {}".format(MAX_END)
)
PARSER.add_option(
  "--video",
  type="float",
  dest="video",
  default=100,
  help="Frequency of averaging the result over a period of time in hertz. From 0.1 to 6400. Default is 100"
)
PARSER.add_option(
  "--datafreq",
  type="float",
  dest="datafreq",
  default=10,
  help="Frequency of averaging the result over a period of time in hertz. From 1 to 20. Default is 10"
)
PARSER.add_option(
  "--rbw",
  type="choice",
  choices=RBW,
  default=str(Rbw.Hz400),
  dest="rbw",
  help="Set rbw for read dommand: %s. Default: %s" % (', '.join(RBW), Rbw.Hz400)
)
PARSER.add_option(
  "--rbwsweep",
  type="choice",
  choices=RBW_SWEEP,
  default=str(RbwSweep.Hz6400),
  dest="rbwsweep",
  help="Set rbw for sweep command: %s. Default: %s" % (', '.join(RBW_SWEEP), RbwSweep.Hz6400)
)
PARSER.add_option(
  "--atten",
  type="choice",
  choices=ATTEN,
  default=str(Attenuation.Db0),
  dest="atten",
  help="Set attenuation: %s. Default: %s" % (', '.join(ATTEN), Attenuation.Db0)
)


def main(argv, options):
    """Entry point."""
    print("Satis read utility v.{}.".format(VERSION))
    if len(argv) != 1:
        PARSER.print_usage()
        return 1

    cmd = argv[0]
    if cmd not in COMMANDS:
        PARSER.print_usage()
        return 1

    socket = websocket.WebSocket()
    socket.connect("ws://{}:8080".format(options.address))
    COMMANDS[cmd](socket, options)
    socket.close()
    return 0


if __name__ == '__main__':  # pragma: no cover
    OPTS, ARGS = PARSER.parse_args()
    sys.exit(main(ARGS, OPTS))
